src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java

```diff
@@ -0,0 +1,330 @@
+
+                try (SAMRecordIterator query = reader.queryOverlapping(shard.interval.getContig(), shard.interval.getStart(), shard.interval.getEnd())) {
+                    ArrayList<GATKRead> readsSoFar = new ArrayList<>();
+
+                    while (query.hasNext()) {
+                        SAMRecord r = query.next();
+                        if (!accept(r, shard.interval)) continue;
+                        if (r.getReadUnmappedFlag()) {
+                            // error out if we accept a read that sticks out too far
+                            // (margin was too tight, the shard may end up missing relevant variants)
+                            if (r.getAlignmentEnd()>lastValidPos) {
+                                throw new GATKException("Margin was too tight, a read sticks out by "+(r.getAlignmentEnd()-shard.interval.getEnd())+", going all the way to "+r.getAlignmentEnd());
+                            }
+                            if (r.getAlignmentStart()<firstValidPos) {
+                                throw new GATKException("Margin was too tight, a read starts early by "+(shard.interval.getStart()-r.getAlignmentStart())+", starting at "+r.getAlignmentStart());
+                            }
+                        }
+
+                        while (currentSubShard.getEnd() < r.getStart()) {
+                            if (!readsSoFar.isEmpty()) {
+                                // ship this one.
+                                ContextShard ret = shard.split(currentSubShard).withReads(readsSoFar);
+                                c.output(ret);
+                                readsSoFar = new ArrayList<>();
+                            }
+                            // move to the next shard
+                            currentSubShard = subshards.get(++currentSubShardIndex);
+                        }
+                        SAMRecordToGATKReadAdapter g = new SAMRecordToGATKReadAdapter(r);
+                        if (null!=optFilter) {
+                            // skip reads that don't pass the filter
+                            if (!optFilter.test(g)) continue;
+                        }
+                        // the header slows serialization too much
+                        g.setHeader(null);
+                        readsSoFar.add(g);
+                    }
+                    // done reading, ship what we have
+                    if (!readsSoFar.isEmpty()) {
+                        ContextShard ret = shard.split(currentSubShard).withReads(readsSoFar);
+                        c.output(ret);
+                    }
+                }
+            }
+
+            @Override
+            public void finishBundle(Context c) throws Exception {
+                if (null!=reader) {
+                    reader.close();
+                }
+                super.finishBundle(c);
+            }
+
+            private boolean accept(SAMRecord r, SimpleInterval region) {
```

Comment:
Recommend that you merge `accept()` with the test against `optFilter`, and make accept() operate on a `GATKRead` by wrapping in the adapter a bit earlier (this will simplify the method quite a bit, since `GATKRead.isUnmapped()` does all of the unmapped checks below for you).